### PR TITLE
DOC: Document release procedure with a walkthrough.

### PR DESCRIPTION
**Pull request by charris**
This is a walkthrough of a release. I have found it helpful to have the commands laid out in a step by step way so that I can go through line by line. This is preliminary, but does show parts of the procedure that could be further automated and provides a place for comments as well as a handy checklist. It is intended to supplement the expanded content in `doc/HOWTO_RELEASE`.

I'm hoping to foist some releases off on other folks, this should help.

### CODE REVIEW

**Comment by charris** [comment]
@rgommers @ev-br Comments would be helpful. I'm trying to make the actual release process easier to follow.

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Maybe add a short comment that ``sdist`` starts with ``git clean -xdf``, so this is OK (doesn't pick up the build artifacts from the last step)

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: I think we should move away from SourceForge, so this is not needed anymore then. Now that we don't distribute .exe installers there's really no use for SF anymore, and it has a pretty bad reputation by now (deservedly).

What is missing is the procedure to upload pdf and html docs to docs.scipy.org, which would be good to add instead.

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Just a note (text is fine): PyPI will likely lose support for signing soon, so this will become simpler.

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: should be ``cd ../terryfy``?

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: same here, ``cd ../numpy``

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: add a comment ``# edit the News section``

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: ``numpy-discussion`` not ``numpy-devel``. also, add ``python-announce``

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: ``cd ../scipy.org``

**Review by rgommers — CHANGES_REQUESTED** [review]
Thanks Chuck, looks quite useful. My main comments are about the missing step for uploading docs to docs.scipy.org and moving away from SourceForge.

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (charris) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: I'll propose dropping sourceforge on the list. We will need to edit some documentation and put a note up on sourceforge. I'll release 1.11.2 there as 1.11.2rc1 is already up.

I don't have permission to upload to docs.scipy.org, how should we deal with that?

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (charris) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Hmm, will there be anything to replace signing, or will it rely on permissions?

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (charris) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: I append `.git` to git repo directories. I can change that for this documentation. This walkthrough probably needs the layout it is written for documented at the beginning.

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (charris) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Will do.

**Comment by charris** [comment]
@rgommers Looks like you have stuck yourself with some work by requesting changes, as they are made but apparently you need to sign off on them. Or maybe github can't deal with rewriting history.

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: not sure yet. there was a long discussion about how signing was not very secure in practice anyway. haven't followed the details, but IIRC it'll go away with the switch from the current infrastructure to Warehouse

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: we should get you access, but for now you can ping me to do the uploads. we're having a bit of an issue with that server at the moment; all the admins had misplaced their passwords. I have ssh access but no admin rights, so can't add you.

Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Instructions are:

    $ make dist    # check that html and pdf docs look right
    $ make upload USERNAME=<username>  # username with ssh access to docs.scipy.org

**Review by rgommers — APPROVED** [review]

**Comment by rgommers** [comment]
> Looks like you have stuck yourself with some work by requesting changes, as they are made but apparently you need to sign off on them

Hit the button. Looks like a PR can be merged without that approval (which is good, otherwise there's a bottleneck on 1 person), it's just the color stays red until the reviewer who requested changes approved them.

Still getting used to the review functionality, but overall i like it.

**Comment by rgommers** [comment]
> Still getting used to the review functionality, but overall i like it.

except that now comments that have been addressed aren't hidden on this conversation page....

**Comment by charris** [comment]
Could get ugly with comments...

I'm not ready to merge this yet, going to add a script for the announcement and a short list of dependencies. At some point should probably cover making the maintenance branch also. Trying to keep text at a minimum though.

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Sorry, that's incorrect. Should be 

    $ make dist    # check that html and pdf docs look right
    $ make upload USERNAME=<username> RELEASE=1.11.2  # username with ssh access to docs.scipy.org

**Comment by ev-br** [comment]
This looks useful!

I second the suggestion to drop Sourceforge in favor of GH releases. 

Here's my rough draft of a checklist for the last scipy release: https://gist.github.com/ev-br/dd5878cb66aa52e8a548fa7eece55fec
The main difference seems to be that I checksum the wheels locally before uploading them to PyPI. The checksums end up in the README file on GH and in the email announcements.  (Also cf some discussion in https://github.com/MacPython/terryfy/issues/6)

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (charris) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Should be possible to put this into paver?

Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: yep not difficult. The ``USERNAME=<username>`` still needs to be input also then, so won't get much shorter.

**Comment by rgommers** [comment]
good point @ev-br, +1 for wheel checksums

**Comment by charris** [comment]
I'll revisit this after the 1.12.0 release. There appear to be two unresolved issues remaining

- Signing/checksumming PyPI uploads
- Documentation

Adding the checksums to the `README.md` could get cumbersome after a while, and it isn't the easiest thing to find. I suppose if the files aren't signed the checksum is needed to check the integrity of the upload.

**Comment by rgommers** [comment]
Just found another issue: tarballs need to be uploaded to GitHub Releases. Now we get the automatically created tarballs from the release tags, but those are not the same (no generated .c files from cython sources for example).

**Comment by charris** [comment]
@rgommers How does one upload tarballs to GitHub Releases?

**Comment by rgommers** [comment]
When you go to https://github.com/numpy/numpy/releases, there's a "draft a new release" button (right top). Click that, then you can edit the description (copy from release notes) and attached files in the UI.

**Comment by charris** [comment]
@rgommers Thanks. Has the password for documentation upload been found? Also, what documentation gets uploaded, HTML and PDF?

**Comment by charris** [comment]
@ev-br I don't see the checksums in the README.md on GH. Also, what do you use for the checksum, `shasum -a 256`, `md5sum`, ?

**Comment by ev-br** [comment]
Other README, here: https://github.com/scipy/scipy/releases
It"s generated by `paver sdist`. When the wheels are ready, I copy them to release/installers folder. Then `paver write_release_and_log` checksums all files in that folder. 
This is scipy's pavement.py file, not sure how different numpy's is.

**Comment by charris** [comment]
@ev-br Do you find the Changelog useful? I'm thinking of dropping it in favor of the PR links. I note the `Source code` links. Where do they come from?

**Comment by rgommers** [comment]
> Has the password for documentation upload been found?

Yep, @pv has root access and I have ssh access again.

> Also, what documentation gets uploaded, HTML and PDF?

Yes, those two. See ``make dist``` and ``make upload`` in ``doc/Makefile``.

**Comment by rgommers** [comment]
> This is scipy's pavement.py file, not sure how different numpy's is.

They're pretty much the same, and differences are probably unintentional.

**Comment by ev-br** [comment]
Source code links are I believe the useless autogenerated ones that Ralf mentioned. The tar.gz, zip and xz files are uploaded manuallyas binaries  instead.
Changelog, no, I've no uses for it. I always thought it's something downstream integrators use for ... I dunno what :-).

**Comment by rgommers** [comment]
Changelog I think is just historically the way people did things. No one really looks at it AFAIK.

**Comment by charris** [comment]
@rgommers So should I get access or should I ping you for that?

**Comment by rgommers** [comment]
I'm happy to do the uploads (ping me when needed), but getting you access may also make sense. Send @pv your ssh key I guess.

**Comment by charris** [comment]
@rgommers Do we upload documents for pre-releases?

**Comment by rgommers** [comment]
no we don't

**Comment by charris** [comment]
@rgommers Could you document the changes needed in the `index.rst` file of `docs.scipy.org`? Also, would it be better to make that a normal pull request as opposed to simply pushing the result?

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: remove ``rc1``, we don't upload pre-release docs

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (rgommers) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Send a PR to https://github.com/scipy/docs.scipy.org with the changes that add links to the new docs. See https://github.com/scipy/docs.scipy.org/pull/7 for an example

File: doc/RELEASE_WALKTHROUGH.rst.txt
Comment (ev-br) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: IIRC twine -s signs them all, e.g. `$ twine -s *macosx*whl` only asks for the passphrase once.

Comment (charris) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: I thought that also, but vaguely recall it not working. Probably need to test again. In any case, `wheel-uploader` doesn't do that and when I asked @matthew-brett  about it he said they could not get it to work.

Comment (charris) [inline doc/RELEASE_WALKTHROUGH.rst.txt]: Might be a function of the twine version. I didn't realize that pip did not automatically upgrade, so was running version 1.6.5.

**Comment by charris** [comment]
I'm going to merge this now for reference. The documentation section will need revision if/when we move the documents to github.

**Pull request merged**
